Packet++/header/IcmpV6Layer.h

```diff
@@ -0,0 +1,300 @@
+
+	/**
+	 * A static method that determines the ICMPv6 protocol type of ICMPv6 layer raw data by looking at the
+	 * icmpv6hdr#type field
+	 * @param[in] data The pointer to the beginning of an ICMPv6 byte stream
+	 * @param[in] dataLen The length of the byte stream
+	 * @return The specific ProtocolType or UnknownProtocol if it can not be determined
+	 */
+	static ProtocolType getIcmpv6Version(uint8_t *data, size_t dataLen);
+
+	/**
+	 * @param[in] type Type to check
+	 * @return True if the layer if of the given type, false otherwise
+	 */
+	bool isMessageOfType(ICMPv6MessageType type) const { return getMessageType() == type; }
+
+	/**
+	 * @return Get the ICMPv6 Message Type
+	 */
+	ICMPv6MessageType getMessageType() const;
+
+	/**
+	 * @return Get the code header field
+	 */
+	uint8_t getCode() const;
+
+	/**
+	 * @return Get the checksum header field in host representation
+	 */
+	uint16_t getChecksum() const;
+
+	/**
+	 * Identifies next layer and tries to create it
+	 */
+	void parseNextLayer();
+
+	/**
+	 * @return The length of the ICMPv6 header
+	 */
+	size_t getHeaderLen() const;
+
+	/**
+	 * Calculate ICMPv6 checksum field
+	 */
+	void computeCalculateFields();
+
+	OsiModelLayer getOsiModelLayer() const { return OsiModelNetworkLayer; }
+
+  protected:
+	IcmpV6Layer() = default;
+
+	IcmpV6Layer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet)
+		: Layer(data, dataLen, prevLayer, packet) {}
+
+  private:
+	void calculateChecksum();
+	icmpv6hdr *getIcmpv6Header() const { return (icmpv6hdr *)m_Data; }
+};
+
+/**
+ * @class ICMPv6EchoRequestLayer
+ * Represents an ICMPv6 echo request protocol layer
+ */
+class ICMPv6EchoRequestLayer : public IcmpV6Layer
+{
+  public:
+	/**
+	 * A constructor that creates the layer from an existing packet raw data
+	 * @param[in] data A pointer to the raw data
+	 * @param[in] dataLen Size of the data in bytes
+	 * @param[in] prevLayer A pointer to the previous layer
+	 * @param[in] packet A pointer to the Packet instance where layer will be stored in
+	 */
+	ICMPv6EchoRequestLayer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet)
+		: IcmpV6Layer(data, dataLen, prevLayer, packet)
+	{
+		m_Protocol = ICMPv6EchoRequest;
+	}
+
+	/**
+	 * A constructor for a new echo request layer
+	 * @param[in] id Echo request identifier
+	 * @param[in] sequence Echo request sequence
+	 * @param[in] data A pointer to echo request payload to set
+	 * @param[in] dataLen The length of the echo request payload
+	 */
+
+	ICMPv6EchoRequestLayer(uint16_t id, uint16_t sequence, const uint8_t *data, size_t dataLen);
+
+	virtual ~ICMPv6EchoRequestLayer() {}
+
+	/**
+	 * @return ICMP echo request data.
+	 */
+	icmpv6_echo_request *getEchoRequestData();
+
+	std::string toString() const;
+
+  private:
+	icmpv6_echo_request m_EchoData;
+};
+
+/**
+ * @class ICMPv6EchoReplyLayer
+ * Represents an ICMPv6 echo reply protocol layer
+ */
+class ICMPv6EchoReplyLayer : public IcmpV6Layer

```

Comment:
If we decide to remove `ICMPv6EchoReply` and `ICMPv6EchoRequest` then both `ICMPv6EchoReplyLayer` and `ICMPv6EchoRequestLayer` will be pretty similar. The only difference will be in the c'tor that sets a different ICMPv6 message type when building a new layer...